components/scream/src/physics/p3/tests/p3_unit_tests.cpp

```diff
@@ -80,6 +81,474 @@ struct UnitWrap::UnitTest<D>::TestP3Func
+
+  }
+
+ 
+  static void ice_water_conservation_tests_device(){
+    using KTH = KokkosTypes<HostDevice>;
+
+    IceWaterConservationData iwdc[1] = {{sp(1e-5), 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, sp(1.1), sp(1e-4), 0.0}};
+
+    // Sync to device
+    KTH::view_1d<IceWaterConservationData> iwdc_host("iwdc_host", 1);
+    view_1d<IceWaterConservationData> iwdc_device("iwdc_host", 1);
+
+    // This copy only copies the input variables.
+    std::copy(&iwdc[0], &iwdc[0] + 1, iwdc_host.data());
+    Kokkos::deep_copy(iwdc_device, iwdc_host);
+
+    // Run the lookup from a kernel and copy results back to host
+    Kokkos::parallel_for(RangePolicy(0, 1), KOKKOS_LAMBDA(const Int& i) {
+      Spack qitot(iwdc_device(0).qitot);
+      Spack qidep(iwdc_device(0).qidep);
+      Spack qinuc(iwdc_device(0).qinuc);
+      Spack qrcol(iwdc_device(0).qrcol);
+      Spack qccol(iwdc_device(0).qccol);
+      Spack qrheti(iwdc_device(0).qrheti);
+      Spack qcheti(iwdc_device(0).qcheti);
+      Spack qiberg(iwdc_device(0).qiberg);
+      Spack qisub(iwdc_device(0).qisub);
+      Spack qimlt(iwdc_device(0).qimlt);
+
+      Functions::ice_water_conservation(qitot, qidep, qinuc, qrcol, qccol, qrheti, qcheti, qiberg, iwdc_device(0).dt, qisub, qimlt);
+
+      iwdc_device(0).qitot = qitot[0];
+      iwdc_device(0).qidep = qidep[0];
+      iwdc_device(0).qinuc = qinuc[0];
+      iwdc_device(0).qrcol = qrcol[0];
+      iwdc_device(0).qccol = qccol[0];
+      iwdc_device(0).qrheti = qrheti[0];
+      iwdc_device(0).qcheti = qcheti[0];
+      iwdc_device(0).qiberg = qiberg[0];
+      iwdc_device(0).qisub = qisub[0];
+      iwdc_device(0).qimlt = qimlt[0];
+    });
+
+  }
+
+  static void run()
+  {
+
+    cloud_water_conservation_tests_device();
+
+    rain_water_conservation_tests_device();
+
+    ice_water_conservation_tests_device();
+
+  }
+
+  static void cloud_water_conservation_unit_bfb_tests(){
+
+    using KTH = KokkosTypes<HostDevice>;
+
+    static constexpr Int max_pack_size = 16;
+    REQUIRE(Spack::n <= max_pack_size);
+
+    CloudWaterConservationData cwdc[max_pack_size] = {
+      {9.9999999999999995e-7, 0.0, 1800.0, 1.5832574016248739e-12, 1.0630996907148179e-12, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+      {6.4285714285714288e-5, 0.0, 1800.0, 5.0577951315583066e-7, 7.7585489624948031e-4, 1.5683327213659326E-4, 1.2893174331809564e-14, 0.0, 5.0463073442953805e-6, 0.0, 5.1387602886199180e-7},
+      {0.0, 0.0, 1800.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+      {7.1428571428571434e-5, 0.0, 1800.0, 5.1480988828550771e-7, 7.7585489624948031e-4, 1.5597668529004373e-4, 4.9926620576534573e-14, 0.0, 6.7718890050008472e-6, 0.0, 7.1052455549903861e-7},
+
+      {9.9999999999999995e-7, 0.0, 1800.0, 1.5832574016248739e-12, 1.0630996907148179e-12, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
+      {6.4285714285714288e-5, 0.0, 1800.0, 5.0577951315583066e-7, 7.7585489624948031e-4, 1.5683327213659326E-4, 1.2893174331809564e-14, 0.0, 5.0463073442953805e-6, 0.0, 5.1387602886199180e-7},
```

Comment:
It's hard to tell at a glance, but I hope that these input were chosen such that different mask values are exercised.

Reply:
Added relevant print statements and verified that all masks were exercised using `ctest -V -R p3_tests_ut_np1_omp1`.